File: WakeUp Sample (Obj-C)/WakeUp Sample/ViewController.m
//
//  ViewController.m
//  WakeUp Sample
//
//  Created by Paul Schmitt on 11/17/16.
//  Copyright © 2016 PathSense. All rights reserved.
//

#import <PSLocation/PSLocation.h>

#import "ViewController.h"

@interface ViewController () <UINavigationBarDelegate, PSLocationManagerDelegate, MKMapViewDelegate>

@property (nonatomic) BOOL preparingMap;
@property (nonatomic, readonly) PSLocationManager *locationManager;

- (void)handleButton:(id)sender;

@end

@implementation ViewController
//----------------------------------------------------------------------------------
- (void)viewDidLoad
{
    [super viewDidLoad];
	_preparingMap = YES;

	[self startLocationManager];
    
    UIBarButtonItem *barButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Set" style:UIBarButtonItemStylePlain target:self action:@selector(handleButton:)];
    [[_navigationBar topItem] setLeftBarButtonItem:barButtonItem animated:YES];
    [barButtonItem setEnabled:NO];
}
//----------------------------------------------------------------------------------
- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    
    CGRect bounds;
    if (@available(iOS 11.0, *)) {
        bounds = [[[self view] safeAreaLayoutGuide] layoutFrame];
    } else {
        bounds = [[self view] bounds];
        bounds.origin.y = [[UIApplication sharedApplication] statusBarFrame].size.height;
        bounds.size.height -= bounds.origin.y;
    }
    CGRect r = bounds;

    CGRect nr = [_navigationBar frame];
    nr.origin.y = r.origin.y;
    [_navigationBar setFrame:nr];
    
    CGRect mr = [_mapView frame];
    mr.origin.y = nr.origin.y + nr.size.height;
    mr.size.height = [[self view] bounds].size.height - mr.origin.y;
    [_mapView setFrame:mr];
}
//----------------------------------------------------------------------------------
- (void)handleButton:(id)sender
{
	#warning set the location you want to monitor departures for here
    [_locationManager setDepartureCoordinate:CLLocationCoordinate2DMake(33.02280304, -117.28318958)];
}

#pragma mark -
//----------------------------------------------------------------------------------
- (void) startLocationManager
{
  	if (_locationManager == nil) {
    	_locationManager = [PSLocationManager new];
        [_locationManager setDelegate:self];
        if ([_locationManager respondsToSelector:@selector(requestAlwaysAuthorization)]) { // iOS 8 or later
            [_locationManager requestAlwaysAuthorization];
        }
        
        if ([_locationManager respondsToSelector:@selector(allowsBackgroundLocationUpdates)]) {  // iOS 9 or later
             [_locationManager setAllowsBackgroundLocationUpdates:YES];
        }
        
        [_locationManager startMonitoringDeparture];
    }
}

#pragma mark - UINavigationBarDelegate
#pragma mark -
//----------------------------------------------------------------------------------
- (UIBarPosition)positionForBar:(id<UIBarPositioning>)bar
{
	return UIBarPositionTopAttached;
}

#pragma mark - PSLocationManagerDelegate
#pragma mark -
//----------------------------------------------------------------------------------
- (void)locationManager:(PSLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status
{
	if (status == kCLAuthorizationStatusNotDetermined) {
    
    } else if (status == kCLAuthorizationStatusRestricted || status == kCLAuthorizationStatusDenied) {
    	NSString *title = NSLocalizedString(@"Location Autorization", nil);
    	NSString *message = NSLocalizedString(@"This application is not authorized to use location services!", nil);

    	UIAlertController *alertController = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
    	[alertController addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"OK", nil) style:UIAlertActionStyleCancel handler:nil]];

    	UIAlertAction *action = [UIAlertAction actionWithTitle:NSLocalizedString(@"Settings...", nil) style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
			NSURL *url = [NSURL URLWithString:UIApplicationOpenSettingsURLString];
			[[UIApplication sharedApplication] openURL:url options:@{} completionHandler:^(BOOL success) {
            }];
    	}];
    	[alertController addAction:action];
    	[self presentViewController:alertController animated:YES completion:^{
        }];
        
    } else {
    }
}
//----------------------------------------------------------------------------------
-(void)locationManager:(PSLocationManager *)manager didUpdateLocations:(NSArray *)locations
{
    // some of these locations may be stale so you will need to filter them as you want
    
    for (CLLocation *location in locations) {
        MKPointAnnotation *pt = [MKPointAnnotation new];
        [pt setTitle:@"Departure"];
        [pt setCoordinate:[location coordinate]];
        [_mapView addAnnotation:pt]; // this does not persist between runs of the app
        [_mapView setCenterCoordinate:[location coordinate]]; // this does not persist between runs of the app
    }
}
//----------------------------------------------------------------------------------
- (void)psLocationManager:(PSLocationManager *)manager didUpdateDepartureCoordinate:(CLLocationCoordinate2D)coordinate
{
	// this will be called whenever you call setDepartureCoordinate
    
    if (_preparingMap) {
    	return;
    }
    
    [_mapView removeAnnotations:[_mapView annotations]];
    MKPointAnnotation *pt = [MKPointAnnotation new];
    [pt setTitle:@"SetLocation"];
    [pt setCoordinate:coordinate];
    [_mapView addAnnotation:pt]; // this does not persist between runs of the app

    MKCoordinateSpan span = MKCoordinateSpanMake(0, 360/pow(2, 16)*[_mapView frame].size.width/256);
    [_mapView setRegion:MKCoordinateRegionMake(coordinate, span) animated:YES];
}
//----------------------------------------------------------------------------------
- (void)psLocationManager:(PSLocationManager *)manager didDepartCoordinate:(CLLocationCoordinate2D)coordinate
{
    // this will be called when a departure is detected -- at this point you need to start getting locations
    // the coordinate passed in will be the coordinate that was passed to setDepartureCoordinate
	
    [manager requestLocation];
}
//----------------------------------------------------------------------------------
- (void)psLocationManagerDepartureMonitoringEnded:(PSLocationManager *)manager
{
	// this will be called once the departure coordinate is no longer being monitored
}

#pragma mark - MKMapViewDelegate
#pragma mark -
//----------------------------------------------------------------------------------
- (void)mapViewDidFinishRenderingMap:(MKMapView *)mapView fullyRendered:(BOOL)fullyRendered
{
    if (_preparingMap) {
    	[[[_navigationBar topItem] leftBarButtonItem] setEnabled:YES];
    }
    _preparingMap = NO;
}
//----------------------------------------------------------------------------------
- (MKAnnotationView *)mapView:(MKMapView *)mapView viewForAnnotation:(id <MKAnnotation>)annotation
{
    MKAnnotationView *result = nil;
    if ([annotation isMemberOfClass:[MKPointAnnotation class]] == YES) {
    	if ([[annotation title] isEqualToString:@"SetLocation"]) {
        	result = [[MKPinAnnotationView alloc] init];
            [(MKPinAnnotationView *)result setPinTintColor:[MKPinAnnotationView redPinColor]];

        } else if ([[annotation title] isEqualToString:@"Departure"]) {
        	result = [[MKPinAnnotationView alloc] init];
            [(MKPinAnnotationView *)result setPinTintColor:[MKPinAnnotationView greenPinColor]];
        }
    }
    return  result;
}

@end

File: WakeUp Sample (Swift)/WakeUp Sample (Swift)/ViewController.swift
//
//  ViewController.swift
//  WakeUp Sample (Swift)
//
//  Created by Paul Schmitt on 12/14/16.
//  Copyright © 2016 PathSense. All rights reserved.
//

import UIKit
import MapKit
import PSLocation

class ViewController: UIViewController, UINavigationBarDelegate, PSLocationManagerDelegate, MKMapViewDelegate
{
	private var preparingMap = true
	private var locationManager : PSLocationManager? = nil
    
	@IBOutlet weak var mapView : MKMapView?
	@IBOutlet weak var navigationBar : UINavigationBar?

    //----------------------------------------------------------------------------------
    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        startLocationManager()
        
        let barButtonItem = UIBarButtonItem.init(title: "Set", style: .plain, target: self, action: #selector(handleButton(sender:)))
        navigationBar?.topItem?.leftBarButtonItem = barButtonItem
        barButtonItem.isEnabled = false
    }
    
    // MARK: -
    //----------------------------------------------------------------------------------
    func startLocationManager()
    {
        guard self.locationManager == nil else {
            return
        }
        let locationManager = PSLocationManager()
        locationManager.setDelegate(self)
        locationManager.requestAlwaysAuthorization()
        locationManager.allowsBackgroundLocationUpdates = true
        locationManager.startMonitoringDeparture()
        self.locationManager = locationManager
    }
    //----------------------------------------------------------------------------------
	@objc func handleButton(sender : UIBarButtonItem)
    {
    	#warning("Set the location you want to monitor departures for here")
    	locationManager?.setDepartureCoordinate(CLLocationCoordinate2DMake(33.02280304, -117.28318958))
    }
    
    // MARK: - UINavigationBarDelegate
    // MARK: -
    //----------------------------------------------------------------------------------
	func position(for bar: UIBarPositioning) -> UIBarPosition
    {
        return .topAttached
    }
    
    // MARK: - PSLocationManagerDelegate
    // MARK: -
    //----------------------------------------------------------------------------------
    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus)
    {
    	if status == .notDetermined {
        
        } else if status == .restricted || status == .denied {
        	
            let alertController = UIAlertController(title: "Location Autorization", message: "This application is not authorized to use location services!", preferredStyle: .alert)
            alertController.addAction(UIAlertAction(title: "OK", style: .cancel, handler: nil))
            
            let action = UIAlertAction(title: "Settings...", style: .default, handler: { (UIAlertAction) in
                if let url = URL(string: UIApplication.openSettingsURLString) {
                    UIApplication.shared.open(url, options: [:], completionHandler: nil)
                }
            })
            alertController.addAction(action)
            present(alertController, animated: true, completion: nil)
            
        } else {
        
        }
	}
    //----------------------------------------------------------------------------------
    func psLocationManager(_ manager: PSLocationManager!, didUpdateDepartureCoordinate coordinate: CLLocationCoordinate2D)
    {
		// this will be called whenever you call setDepartureCoordinate
		
		guard let mapView = mapView else {
            return
        }

    	mapView.removeAnnotations(mapView.annotations)
        let pt = MKPointAnnotation.init()
        pt.title = "SetLocation"
        pt.coordinate = coordinate
        mapView.addAnnotation(pt)
        
        let span = MKCoordinateSpan.init(latitudeDelta: 0, longitudeDelta: (360.0/pow(2.0, 16.0)) * (Double(mapView.frame.size.width)/256.0))
        mapView.setRegion(MKCoordinateRegion.init(center: coordinate, span: span), animated:true)
	}
    //----------------------------------------------------------------------------------
    func psLocationManager(_ manager: PSLocationManager!, didDepart coordinate: CLLocationCoordinate2D, at location: CLLocation!)
    {
		let pt = MKPointAnnotation.init()
		pt.title = "Departure"
		pt.coordinate = location.coordinate
		mapView?.addAnnotation(pt)
    }
    //----------------------------------------------------------------------------------
    func psLocationManagerDepartureMonitoringEnded(_ manager: PSLocationManager!)
    {
		// this will be called once the departure coordinate is no longer being monitored
	}

    // MARK: - MKMapViewDelegate
    // MARK: -
    //----------------------------------------------------------------------------------
    func mapViewDidFinishLoadingMap(_ mapView: MKMapView)
    {
    	guard preparingMap else {
            return
        }
		
		navigationBar?.topItem?.leftBarButtonItem?.isEnabled = true
        preparingMap = false
	}
    //----------------------------------------------------------------------------------
    func mapView(_ mapView: MKMapView, viewFor annotation: MKAnnotation) -> MKAnnotationView?
    {
    	var result : MKAnnotationView? = nil
        
        guard let annotation = annotation as? MKPointAnnotation else {
            return result
        }
        
        if annotation.title == "SetLocation" {
            result = MKPinAnnotationView()
            result?.tintColor = MKPinAnnotationView.redPinColor()
        } else if annotation.title == "Departure" {
            result = MKPinAnnotationView()
            result?.tintColor = MKPinAnnotationView.greenPinColor()
        }

        return result
	}
}
